Introduce `evicted-at`/`last-evicted` timestamps

evanlinjin opened the pull request:
Fixes #1740.

### Description

This PR replaces #1765. For context and the original discussion that led to this change, please refer to [this comment](https://github.com/bitcoindevkit/bdk/pull/1765#issuecomment-2611491174).

This PR addresses a potential malicious double-spending issue by introducing improvements to unconfirmed transaction tracking. Key changes include the addition of `TxUpdate::missing` that tracks transactions that have been replaced and are no longer in the mempool, and the inclusion of `last_evicted` and `evicted_at` timestamps in `TxGraph` to track when a transaction was last deemed missing. 

`SpkWithExpectedTxids` is introduced in `SpkClient` to track expected `Txid`s for each `spk`. During a sync, if any `Txid`s from `SpkWithExpectedTxids` are not in the current history of an `spk` obtained from the chain source, those `Txid`s are considered missing. Support for `SpkWithExpectedTxids` has been added to both `bdk_electrum` and `bdk_esplora` chain source crates.

The canonicalization algorithm is updated to disregard transactions with a `last_evicted` timestamp greater than or equal to their `last_seen` timestamp, except in cases where transitivity rules apply.

### Changelog notice

* `TxUpdate::missing` tracks transactions that have been replaced and are no longer present in mempool.
* `last_evicted` and `evicted_at` timestamps in `TxGraph` track when a transaction was last marked as missing from the mempool.
* The canonicalization algorithm now disregards transactions with a `last_evicted` timestamp greater than or equal to it's `last_seen` timestamp, except when a canonical descendant exists due to rules of transitivity.
* `SpkWithExpectedTxids` in `SpkClient` keeps track of expected `Txid`s for each `spk`.
* `SpkWithExpectedTxids` support added for `bdk_electrum` and `bdk_esplora`.
* `SyncRequestBuilder::expected_txids_of_spk` adds an association between `Txid` and `spk`.
* `SyncRequestExt::check_unconfirmed_statuses` adds unconfirmed transactions alongside their `spk`s during sync.

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for the new feature
* [x] I've added docs for the new feature

#### Bugfixes:

* [x] This pull request breaks the existing API
* [x] I've added tests to reproduce the issue which are now passing
* [x] I'm linking the issue being fixed by this PR

evanlinjin committed:
feat(core)!: Make `TxUpdate` non-exhaustive

This may, if we introduce new fields to `TxUdpate`, they can be minor
non-breaking updates.

evanlinjin committed:
feat(core): Add `TxUpdate::evicted`

This is a set of txids evicted from the mempool.

evanlinjin committed:
feat(chain)!: Change `TxGraph` to understand evicted-at & last-evicted

The evicted-at and last-evicted timestamp informs the `TxGraph` when the
transaction was last deemed as missing from the mempool.

The canonicalization algorithm is changed to disregard transactions with
a last-missing timestamp greater or equal to it's last-seen timestamp.
The exception is when we have a canonical descendant due to rules of
transitivity.

LagginTimes committed:
feat(core): add `SpkWithExpectedTxids` support for `SpkClient`

evanlinjin committed:
feat(chain): Update rusqlite_impl to persist `last_evicted`

evanlinjin committed:
feat(core): Add `SyncRequestBuilder::expected_txids_of_spk`

This is for conveniently adding associations of txid <-> spk. We expect
that these txids exist in the spk history. Otherwise, it means the tx is
evicted from the mempool and we need to update the `missing_at` value in
the sync response.

LagginTimes committed:
feat(electrum): add `SpkWithExpectedTxids` support

LagginTimes committed:
feat(esplora): add `SpkWithExpectedTxids` support

evanlinjin committed:
feat(chain): Add `SyncRequestExt::check_unconfirmed_statuses`

This is a convenience method for adding unconfirmed txs alongside their
associated spks the the sync request. This way, we will be able to
detect evictions of these transactions from the mempool.

LagginTimes committed:
test(electrum): `detect_receive_tx_cancel`

LagginTimes committed:
test(esplora): `detect_receive_tx_cancel`

notmandatory commented:
Looks like only breaking changes are on `core` and `chain`, so we'll be able to roll this out in a non-breaking 1.x `wallet` release?

evanlinjin commented on crates/chain/src/indexer/keychain_txout.rs:
```diff
@@ -879,6 +881,20 @@ pub trait SyncRequestBuilderExt<K> {
+    ///
+    /// We expect that the chain source should include these txids in their spk histories. If not,
+    /// the transaction has been evicted for some reason and we will inform the receiving
+    /// structures in the response.
```
Update docs here.

evanlinjin commented on crates/core/src/spk_client.rs:
```diff
@@ -86,6 +86,14 @@ impl SyncProgress {
+    pub spk: ScriptBuf,
+    /// Txid history.
+    pub txids: HashSet<Txid>,
+}
```
Maybe `SpkWithExpectedTxids` to be more explicit?

evanlinjin commented on crates/core/src/spk_client.rs:
```diff
@@ -301,6 +369,11 @@ impl<I> SyncRequest<I> {
     }
 
+    /// Iterate over [`ScriptBuf`]s with corresponding [`Txid`] histories contained in this request.
+    pub fn iter_spks_with_history(&mut self) -> impl ExactSizeIterator<Item = SpkWithHistory> + '_ {
```
Rename to `iter_spks_with_expected_txids`

LLFourn commented:
@notmandatory core breaking change is also a wallet crate breaking change unfortunately. It is a minor breaking change (some struct from core is changing that no one inspects directly in their code probably) but it technically is a breaking change so it would have to be in bdk_wallet v2. This is a pretty important improvement so I'd say it justifies a v2 release by itself when it's merged.

LLFourn commented on crates/chain/src/indexer/keychain_txout.rs:
```diff
@@ -879,6 +881,23 @@ pub trait SyncRequestBuilderExt<K> {
 
     /// Add [`Script`](bitcoin::Script)s that are revealed by the `indexer` but currently unused.
     fn unused_spks_from_indexer(self, indexer: &KeychainTxOutIndex<K>) -> Self;
+
```
:pick: I don't really see the attraction of a extension trait for this stuff. I think this should all be just member methods.

LLFourn commented on crates/chain/src/indexer/keychain_txout.rs:
```diff
@@ -892,6 +911,29 @@ impl<K: Clone + Ord + core::fmt::Debug> SyncRequestBuilderExt<K> for SyncRequest
+    fn check_for_missing_txs<A, C>(
+        self,
+        indexer: &KeychainTxOutIndex<K>,
+        canonical_iter: CanonicalIter<A, C>,
```
This seems extremely awkward. What about ChainOracles that are not infallible? Why does it need be an `KeychainTxOutIndex` -- it just calls `inner` in the implementation.

How about you just put this method on the tx graph. Like:

```rust
impl<I, D: AsRef<SpkTxOutIndex<I>>> IndexedTxGraph<D> {
    fn sync_request_add_expected_unconfirmed_txs<O: ChainOracle>(&self, sync_request: &mut SyncRequestBuilder, chain_oracle: &O, range: impl RangeBounds<I>) -> Result<(), O::Error> { ... }
}

impl<K> IndexedTxGraph<KeychainTxOutIndex<K>> {
    fn keychain_sync_request_add_expected_unconfirmed_txs<O: ChainOracle>(&self, sync_request: &mut SyncRequestBuilder, chain_oracle: &O, range: impl RangeBounds<K>) -> Result<(), O::Error> { ... }
}
```

ValuedMammal replied:
Building on this comment I put some ideas for a `IndexedTxGraph` method in evanlinjin/bdk#14

evanlinjin replied:
Thanks @LLFourn, the infallible part was pure laziness.

I wanted to have the extension traits since `bdk_core::spk_client` types have a builder API. However, @ValuedMammal's has already provided a superior solution - I'm happy to go with that!

LLFourn commented on crates/chain/src/indexer/spk_txout.rs:
```diff
@@ -334,4 +334,21 @@ impl<I: Clone + Ord + core::fmt::Debug> SpkTxOutIndex<I> {
         input_matches || output_matches
     }
+
+    /// Find relevant script pubkeys associated with a transaction.
```
:pick: doesn't define relevant here.

LLFourn commented on crates/chain/src/indexer/spk_txout.rs:
```diff
@@ -334,4 +334,21 @@ impl<I: Clone + Ord + core::fmt::Debug> SpkTxOutIndex<I> {
+
+    /// Find relevant script pubkeys associated with a transaction.
+    ///
+    /// The script pubkeys of the transaction's outputs and previous outputs as scanned.
```
unhelpful sentence. Why. What do I use this method for.

Also needs to explain that inputs would not be detected as relevant unless the parents have been scanned.

LLFourn commented on crates/chain/src/tx_graph.rs:
```diff
@@ -715,6 +717,30 @@ impl<A: Anchor> TxGraph<A> {
         changeset
     }
 
+    /// Inserts the given `missing_at` for `txid`
```
:wrench: there is no explanation of `missing_at`.

LLFourn reviewed (changes requested):
ConceptACK. Nice ideas here. I think chain API needs a clean up.

ValuedMammal committed:
feat(chain): add method `iter_spks_with_expected_txids`

ValuedMammal committed:
test(rpc): add `test_expect_tx_evicted`

ValuedMammal commented on crates/chain/src/tx_graph.rs:
```diff
@@ -145,6 +145,7 @@ pub struct TxGraph<A = ConfirmationBlockTime> {
     spends: BTreeMap<OutPoint, HashSet<Txid>>,
     anchors: HashMap<Txid, BTreeSet<A>>,
     last_seen: HashMap<Txid, u64>,
+    last_missing: HashMap<Txid, u64>,
```
I think it would be nice if we had a data structure like this and then `last_seen` would become `HashMap<Txid, LastSeen>`

```rust
#[derive(Debug, Clone, Copy, Default, PartialEq)]
struct LastSeen {
    /// seen at
    seen_at: u64,
    /// evicted at
    evicted_at: Option<u64>,
}
```

evanlinjin replied:
I quite like this idea since there has been talks about adding more timestamps.

I.e. `first-seen` timestamp, which will help with tx ordering when listing transactions.

```rust
struct MempoolTimestamps {
    pub first_seen: u64,
    pub last_seen: u64,
    pub last_evicted: Option<u64>,
}
```

I also like the term `evicted` as it's more descriptive than `missing`.

oleonardolima replied:
I agree, and +1 on this idea, it makes it clear, and easy to "derive" its lifecycle and it's and can be useful info to expose to the users.

LagginTimes replied:
After some discussion with @evanlinjin, it was decided that this was best done in a separate PR as it may be too breaking of a change.

evanlinjin replied:
Yes. On second thought, this introduces an invariant that `last_evicted` must exist only if we have had a `seen_at` value. Because of the existence of this invariant, this invariant must be reflected in `tx_graph::ChangeSet` by changing `ChangeSet::seen_at` to have values of `MempoolTimestamps`. This change breaks our policy of changes to changeset which was to only add new fields.

Edit: Maybe we can make it a user-facing API.

ValuedMammal commented on crates/chain/src/tx_graph.rs:
```diff
@@ -110,19 +110,19 @@ use core::{
+            .flat_map(|(txid, anchors)| anchors.into_iter().map(move |a| (a, txid)))
+            .collect();
+        tx_update.seen_ats = graph.last_seen.into_iter().collect();
+        tx_update
```
It seems like `TxUpdate` would need another field `evicted_at` in order to maintain roundtrip convertibility between `TxUpdate` and `TxGraph`.

evanlinjin replied:
Do you think the convertibility between `TxGraph` and `TxUpdate` is important? I'm thinking maybe we should just get rid of it. One can just construct an empty `TxGraph` and apply an `TxUpdate`.

ValuedMammal replied:
It's probably ok to forego complete fungibility between graphs and updates, but at the same time I wonder how useful are the seen-ats without the corresponding evictions?

evanlinjin replied:
> but at the same time I wonder how useful are the seen-ats without the corresponding evictions?

@ValuedMammal what do you mean by this?

I think it makes sense to completely remove convertibility of `TxGraph -> TxUpdate` and `TxUpdate -> TxGraph`.

ValuedMammal replied:
Just pointing out that we lose information going from `TxGraph` to `TxUpdate`. Since both last-seen and last-evicted can influence canonicalization, it might make sense to include both (or neither).

Maybe it depends on the use case for `TxUpdate`. If updating a tx graph from a `spk_client` based chain source, the `seen_ats` don't seem to be as important compared to treating graphs and updates as interchangeable. If we remove convertibility, would it make sense to change this `From` impl to a method on `TxGraph` that returns a `TxUpdate`? We could potentially develop different kinds of updates, like one that is only relevant to a subset of keychains.

ValuedMammal commented on crates/chain/src/tx_graph.rs:
```diff
@@ -715,6 +717,30 @@ impl<A: Anchor> TxGraph<A> {
     }
 
+    /// Inserts the given `missing_at` for `txid`
+    pub fn insert_missing_at(&mut self, txid: Txid, missing_at: u64) -> ChangeSet<A> {
```
`insert_evicted_at`?

```suggestion
    pub fn insert_evicted_at(&mut self, txid: Txid, evicted_at: u64) -> ChangeSet<A> {
```

ValuedMammal commented on crates/chain/src/tx_graph.rs:
```diff
@@ -1139,6 +1182,8 @@ pub struct ChangeSet<A = ()> {
     /// Added last-seen unix timestamps of transactions.
     pub last_seen: BTreeMap<Txid, u64>,
+    /// Added timestamps of when a transaction is last missing from the mempool.
+    pub last_missing: BTreeMap<Txid, u64>,
```
`last_evicted`?
```suggestion
    pub last_evicted: BTreeMap<Txid, u64>,
```

oleonardolima commented on crates/chain/src/tx_graph.rs:
```diff
@@ -969,9 +1007,14 @@ impl<A: Anchor> TxGraph<A> {
-    /// Transactions without last-seens are excluded.
-    pub fn txids_by_descending_last_seen(&self) -> impl ExactSizeIterator<Item = (u64, Txid)> + '_ {
-        self.txs_by_last_seen.iter().copied().rev()
+    /// Transactions without last-seens are excluded. Transactions with a last-missing timestamp
```
``` suggestion
    /// Transactions without last-seens are excluded. Transactions with a last-evicted timestamp
```

oleonardolima commented on crates/chain/src/tx_graph.rs:
```diff
@@ -1139,6 +1182,8 @@ pub struct ChangeSet<A = ()> {
     pub anchors: BTreeSet<(A, Txid)>,
     /// Added last-seen unix timestamps of transactions.
     pub last_seen: BTreeMap<Txid, u64>,
+    /// Added timestamps of when a transaction is last missing from the mempool.
```
``` suggestion
    /// Added timestamps of when a transaction is last evicted from the mempool.
```

oleonardolima reviewed (commented):
A general ConceptACK, some small nits.

oleonardolima commented on crates/chain/src/tx_graph.rs:
```diff
@@ -765,6 +791,14 @@ impl<A: Anchor> TxGraph<A> {
                 changeset.merge(self.insert_seen_at(txid, seen_at));
             }
         }
+        for txid in update.missing {
```
I think it'd help to update `apply_update_at` documentation in order to mention the `evicted_at` too, in [lines](https://github.com/bitcoindevkit/bdk/pull/1811/files#diff-edb82669f839091d04fe6d361bb8cf7ad7f738ba717663cb2735a5bc64f525cfR758-R770).

evanlinjin committed:
feat(chain)!: `IndexedTxGraph::expected_unconfirmed_spk_txids`

Make this method work when the indexer is `KeychainTxOutIndex`. We
reintroduce the ability to get the internal `SpkTxOutIndex` from
`KeychainTxOutIndex` so that `SpkTxOutIndex::relevant_spks_of_tx` is
callable from `KeychainTxOutIndex`.

This commit renames `iter_spks_with_expected_txids` to
`expected_unconfirmed_spk_txids` for `TxGraph`, `IndexedTxGraph` and
`SyncRequestBuilder`. Docs are also improved to explain how these
methods are useful.

Remove unused `SyncRequestBuilder` methods.

evanlinjin commented on crates/bitcoind_rpc/tests/test_emitter.rs:
```diff
@@ -731,3 +731,103 @@ fn no_agreement_point() -> anyhow::Result<()> {
+        .collect::<Vec<_>>();
+    assert_eq!(exp_txids, vec![(txid_1, spk)]);
+    let mempool = emitter
+        .mempool()?
```
I'm wondering whether it will make sense to change the signature of `Emitter::mempool` to take in `expected_unconfirmed_txids` and output `TxUpdate`. This way the caller won't accidentally forget to check `expected_unconfirmed_txids` of the graph.

We can also take the opportunity to keep track of full txs that are already emitted and not re-emit them.

ValuedMammal replied:
I assumed it would be more of an opt-in feature, since it might feel like a significant behavior change to "eagerly evict" a tx the first time it appears missing.

evanlinjin replied:
It would still be an opt-in feature though? If you provide an empty `expected_unconfirmed_txids` input, nothing would be marked as evicted.

ValuedMammal replied:
Oh gotcha. I think that could work.

evanlinjin commented on crates/chain/src/indexer/spk_txout.rs:
```diff
@@ -334,4 +340,23 @@ impl<I: Clone + Ord + core::fmt::Debug> SpkTxOutIndex<I> {
+
+    /// Find relevant script pubkeys associated with a transaction for tracking and validation.
+    ///
+    /// Returns a set of script pubkeys from [`SpkTxOutIndex`] that are pertinent to the outputs and
```
`pertinent` is synonymous with `relevant` right? Can we just use `relevant` to be consistent and it's a much more common term.

evanlinjin committed:
docs(chain): Update `tx_graph` module docs

* Remove duplicate paragraphs about `ChangeSet`s.
* Add "Canonicalization" section which expands on methods that require
  canonicalization and the associated data types used in the
  canonicalization algorithm.

ValuedMammal commented on crates/chain/src/tx_graph.rs:
```diff
@@ -1110,6 +1192,51 @@ impl<A: Anchor> TxGraph<A> {
+
+                core::iter::repeat(txid).zip(spks)
+            })
+    }
```
To make it fallible I allocated a vec for the return value, not sure which way you prefer.

<details>

<summary>expand</summary>

```rust
pub fn expected_unconfirmed_spk_txids<'a, C, I>(
    &'a self,
    chain: &'a C,
    chain_tip: BlockId,
    indexer: &'a impl AsRef<SpkTxOutIndex<I>>,
    range: impl RangeBounds<I> + 'a,
) -> Result<impl Iterator<Item = (Txid, ScriptBuf)> + 'a, C::Error>
where
    C: ChainOracle,
    I: fmt::Debug + Clone + Ord + 'a,
{
    let mut spk_txs = vec![];
    for res in self.try_list_canonical_txs(chain, chain_tip) {
        let canon_tx = res?;
        if canon_tx.chain_position.is_confirmed() {
            continue;
        }
        let txid = canon_tx.tx_node.txid;
        let tx = canon_tx.tx_node.tx;
        let outpoints = tx.input.iter().map(|txin| txin.previous_output).chain(
            tx.output
                .iter()
                .enumerate()
                .map(|(vout, _)| OutPoint::new(txid, vout as u32)),
        );
        for op in outpoints {
            if let Some((index, txo)) = indexer.as_ref().txout(op) {
                if range.contains(index) {
                    spk_txs.push((txid, txo.script_pubkey.clone()));
                }
            }
        }
    }
    Ok(spk_txs.into_iter())
}
```

</details>

LagginTimes committed:
refactor(rpc)!: update `mempool` interface and test code

evanlinjin commented on example-crates/example_bitcoind_rpc_polling/src/main.rs:
```diff
@@ -256,7 +253,22 @@ fn main() -> anyhow::Result<()> {
+                                    (tx, seen_at)
+                                })
+                                .collect();
+                            tx.send(Emission::Mempool(mempool_txs))?;
```
~It will make more sense to change `Emission::Mempool` to contain a `TxUpdate`.~

Edit: `Emission::Mempool` should contain `MempoolEvent` instead. Refer to https://github.com/bitcoindevkit/bdk/pull/1811#discussion_r1956619412

evanlinjin commented on example-crates/example_wallet_rpc/src/main.rs:
```diff
@@ -130,7 +131,21 @@ fn main() -> anyhow::Result<()> {
+                (tx, seen_at)
+            })
+            .collect();
+        sender.send(Emission::Mempool(mempool_txs))?;
```
~Same here. Change `Emission::Mempool` instead.~

Edit: `Emission::Mempool` should contain `MempoolEvent` instead. Refer to https://github.com/bitcoindevkit/bdk/pull/1811#discussion_r1956619412

evanlinjin commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -91,44 +95,51 @@ impl<'c, C: bitcoincore_rpc::RpcApi> Emitter<'c, C> {
+            if !tx_update.seen_ats.contains_key(&expected_txid) {
+                tx_update.evicted.insert(expected_txid);
+            }
+        }
```
This is wrong as we skip including some mempool txs into `TxUpdate::txs` (check line 112).

Maybe create a `HashSet<Txid>` from the `.get_raw_mempool_verbose()` result directly to compare the `expected_unconfirmed_txids` against.

evanlinjin commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -74,8 +74,12 @@ impl<'c, C: bitcoincore_rpc::RpcApi> Emitter<'c, C> {
+    pub fn mempool<A>(
+        &mut self,
+        expected_unconfirmed_txids: impl IntoIterator<Item = Txid>,
+    ) -> Result<TxUpdate<A>, bitcoincore_rpc::Error> {
```
Okay I just realized that returning `TxUpdate` here is a mistake because when we apply `TxUpdate` we don't filter out irrelevant txs.

I think we should do the following:

`mempool()` should return a `MempoolEvent` struct.

```rust
pub struct MempoolEvent {
    pub unseen_txs: Vec<(Transaction, u64)>,
    pub all_txids: HashSet<Txid>,
}
```

We can remove the `expected_unconfirmed_txids` from the `mempool` method as we can use `MempoolEvent::all_txids` to find evicted txs (let's create a helper method on `MempoolEvent` to do this!

evanlinjin replied:
Also the `A` generic is annoying here. Mempool txs wouldn't be anchored.

evanlinjin commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -91,44 +95,49 @@ impl<'c, C: bitcoincore_rpc::RpcApi> Emitter<'c, C> {
-                        // the tx is confirmed or evicted since `get_raw_mempool_verbose`
-                        Err(err) if err.is_not_found_error() => return None,
-                        Err(err) => return Some(Err(err)),
-                    };
```
No need to modify this now. Can you reduce the diff by not changing to a for loop?

evanlinjin commented on crates/bitcoind_rpc/tests/test_emitter.rs:
```diff
@@ -806,23 +813,29 @@ fn test_expect_tx_evicted() -> anyhow::Result<()> {
+    let exp_txids = exp_spk_txids.into_iter().map(|(txid, _)| txid);
+    let evicted_txids = mempool_event.evicted_txids(exp_txids);
+    for txid in evicted_txids {
+        let _ = graph.insert_evicted_at(txid, seen_at + 1);
```
Why do we need the `+1`?

evanlinjin commented on example-crates/example_bitcoind_rpc_polling/src/main.rs:
```diff
@@ -287,7 +284,8 @@ fn main() -> anyhow::Result<()> {
                     Emission::Mempool(mempool_txs) => {
-                        let graph_changeset = graph.batch_insert_relevant_unconfirmed(mempool_txs);
+                        let graph_changeset =
+                            graph.batch_insert_relevant_unconfirmed(mempool_txs.emitted_txs);
```
Could you modify this to check for evicted txs?

evanlinjin commented on crates/chain/src/tx_graph.rs:
```diff
@@ -1161,6 +1163,50 @@ impl<A: Anchor> TxGraph<A> {
+        let chain_tip = chain.get_chain_tip().unwrap();
+
+        self.list_canonical_txs(chain, chain_tip)
+            .filter(|c| !c.chain_position.is_confirmed() && indexer.is_tx_relevant(&c.tx_node))
```
Any confirmed tx that is missing from the Electrum's spk history should also be evicted. Since this means that there is a reorg and the tx is not re-introduced to the mempool (evicted).

evanlinjin commented on crates/chain/src/tx_graph.rs:
```diff
@@ -1164,48 +1163,46 @@ impl<A: Anchor> TxGraph<A> {
+        indexer: &'a impl AsRef<SpkTxOutIndex<I>>,
         range: impl RangeBounds<I> + 'a,
-    ) -> impl Iterator<Item = (Txid, ScriptBuf)> + 'a
+    ) -> Result<impl Iterator<Item = (Txid, ScriptBuf)> + 'a, C::Error>
```
There is no point returning an `impl Iterator` if it's just a `Vec`.

evanlinjin commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -144,6 +154,37 @@ impl<'c, C: bitcoincore_rpc::RpcApi> Emitter<'c, C> {
+/// A new emission from mempool.
+#[derive(Debug)]
+pub struct MempoolEvent {
+    /// Emitted mempool transactions with their first‐seen unix timestamps.
+    pub emitted_txs: Vec<(Transaction, u64)>,
```
```suggestion
    /// Unemitted transactions or transactions with ancestors that are unseen by the receiver.
    ///
    /// To understand the second condition, consider a receiver which filters transactions based on
    /// whether it alters the UTXO set of tracked script pubkeys. If an emitted mempool transaction
    /// spends a tracked UTXO which is confirmed at height `h`, but the receiver has only seen up to
    /// block of height `h-1`, we want to re-emit this transaction until the receiver has seen the
    /// block at height `h`.
    pub new_txs: Vec<(bitcoin::Transaction, u64)>,

```

"Emitted" is not descriptive as the whole `MempoolEvent` struct is emitted. Also I think we need to describe clearer what exactly is contained in this field.

evanlinjin commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -144,6 +154,37 @@ impl<'c, C: bitcoincore_rpc::RpcApi> Emitter<'c, C> {
+    /// Emitted mempool transactions with their first‐seen unix timestamps.
+    pub emitted_txs: Vec<(Transaction, u64)>,
+
+    /// Set of all [`Txid`]s from the raw mempool result, including transactions that may have been
+    /// confirmed or evicted during processing. This is used to determine which expected
+    /// transactions are missing.
+    pub raw_mempool_txids: HashSet<Txid>,
```
```suggestion
    /// Txids of all transactions currently present in the mempool.
    ///
    /// This is useful to determine whether transactions are evicted from the mempool.
    pub all_txids: HashSet<Txid>,
```

I think this is more concise and clearer.

evanlinjin commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -144,6 +154,37 @@ impl<'c, C: bitcoincore_rpc::RpcApi> Emitter<'c, C> {
+    /// transactions are missing.
+    pub raw_mempool_txids: HashSet<Txid>,
+
+    /// The latest first-seen epoch of emitted mempool transactions.
+    pub last_seen: u64,
```
```suggestion
    /// The latest timestamp of when a transaction entered the mempool.
    ///
    /// This is useful for setting the timestamp for evicted transactions.
    pub latest_update_time: u64,
```

What do you think about this?

evanlinjin commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -144,6 +154,37 @@ impl<'c, C: bitcoincore_rpc::RpcApi> Emitter<'c, C> {
+    /// Given an iterator of expected [`Txid`]s, return those that are missing from the mempool.
+    pub fn evicted_txids(
+        &self,
+        expected_unconfirmed_txids: impl IntoIterator<Item = Txid>,
```
For spk-based syncing we want to use all canonical txids (including confirmed) to determine evicted txs. The reason is because spk-based APIs returns both confirmed and unconfirmed under a given spk and that a missing previously-confirmed tx is also evicted from the mempool.

For rpc-based, we obviously want to only input unconfirmed txs as block (confirmed events) are handled separately...

We can add an `include_confirmed: bool` param on `TxGraph`/`IndexedTxGraph` methods that lists expected txs... but that does feel seen like it's easy to screw up.

Another idea is to compare the `MempoolEvent` txids to the previous `MempoolEvent` txids and the difference that does not end up in a block is evicted... However, when we initiate the `Emitter`, we still need the wallet state of unconfirmed txs as we no longer have all the data from the last `MempoolEvent` emission.

evanlinjin replied:
Maybe it's best to combine those two ideas.

This way we don't have to get something from our receiving structures (`expected_unconfirmed_txids`) before applying the update.

ValuedMammal replied:
I'm not sure that the entire txid history is strictly needed in the expected set.

> a missing previously-confirmed tx is also evicted from the mempool.

Perhaps, but maybe we're only concerned with reorgs below some number of confirmations? For a busy wallet the set of expected txids could be quite large which could hinder sync performance over time, considering that the normal fate of a transaction is to gain confirmations and never disappear from the blockchain.

ValuedMammal replied:
> We can add an `include_confirmed: bool` param on `TxGraph`/`IndexedTxGraph` methods that lists expected txs... but that does feel seen like it's easy to screw up.

Or have a method that takes an iterator of canonical txs, that way the caller can filter them on a number of confirmations for example.

evanlinjin commented:
Replaced by #1839
